For #5557 Get Cert Issuer Name

kglazko opened the pull request:
--- 
<!-- Text above this line will be added to the commit once "bors" merges this PR -->

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features 

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

kglazko commented:
This still needs to be tested with a local build of Fenix, do not merge.

pocmo commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -742,6 +742,10 @@ class GeckoEngineSession(
     }
 
+    internal fun GeckoSession.ProgressDelegate.SecurityInformation.getIssuerName(): String? {
+            return certificate?.issuerDN?.name?.substringAfterLast("O=")?.substringBeforeLast(",C=")
```
That is an interesting idea to parse this!

I saw that Android also has a `DName` class. I wonder if we can throw the `String` into that class and get the results back:
https://developer.android.com/reference/kotlin/android/net/http/SslCertificate.DName.html?hl=en

kglazko replied:
Oh my gosh that is amazing, I had no idea that existed. Let me try it!

kglazko commented:
@csadilek @pocmo I added a comment, I couldn't get Sebastian's implementation to work (first AS was complaining and then when I modified it to work, the unit tests failed). Please let me know how I should proceed! I added 3 tests now, please let me know how I should proceed :)

csadilek commented on components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:
```diff
@@ -742,6 +744,12 @@ class GeckoEngineSession(
+    internal fun GeckoSession.ProgressDelegate.SecurityInformation.getIssuerName(): String? {
+        return certificate?.issuerDN?.name?.substringAfterLast("O=")?.substringBeforeLast(",C=")
+        // I can't get the below implementation to work correctly
+        // certificate?.issuerDN?.name?.let { SslCertificate.DName(it).getDName() }
```
This seems to work using:
```
return SslCertificate(certificate).issuedBy.oName
```

You will need to create an instance of `SslCertificate` first to create an instance of `DName`. Tested and this works in the app, but we need to figure out how to mock this correctly for unit testing.

csadilek replied:
At least with the androidx test runner I can't get a unit test running that depends on `android.net.http.SslCertificate.DName`. I am getting `java.lang.NoClassDefFoundError: com/android/org/bouncycastle/asn1/x509/X509Name`... tried for a while to resolve this without much luck.

So, I am OK landing this simple parser for now which we can cover with tests at least. @pocmo wdyt?

pocmo replied:
With the simple parser I am not sure whether we will be guaranteed to always get the string ordered in the same way with all elements always present. If not that could be causing issues. If we want to land this then maybe we should get another review from the GV team to make sure this is okay.

The other option here is to move this specific test to `androidTest` to run it on an actual device/emulator where those classes exist. We already run other `androidTest` tests for `browser-engine-gekco-nightly` on Firebase (and others hopefully soon too). I think I would prefer that. What do you think?

csadilek replied:
@pocmo and I talked about this: Let's land the PR as is with the simple solution for now. 

Otherwise we have to parse the full SSL certificate every time, which is costly. We can address this in https://github.com/mozilla-mobile/android-components/issues/5557 and also talk to GV team again to see if they can provide a better API for us.

@kglazko There are some lint issues to address and would you mind squashing your commits then we can land this :).

kglazko replied:
Thank you! Sounds great.

csadilek commented:
bors r+

csadilek commented:
bors retry

(network failures)
